docs/architecture/adr-073-libp2p.md

```diff
@@ -0,0 +1,189 @@
+`Endpoint`, `Transport`, `Connection`, and `PeerManager` abstractions
+and leave the reactors, `p2p.Router` and `p2p.Channel`
+abstractions.
+
+LibP2P may obviate the need for a dedicated peer exchange (PEX)
+reactor, which would also in turn obviate the need for a dedicated
+seed mode. If this is the case, then all of this functionality would
+be removed.
+
+If it turns out, based on the advice of Protocol Labs, that it makes
+sense to maintain separate pubsub or gossipsub topics
+per-message-type, then the `Router` abstraction may dissolve
+entirely.
+
+## Detailed Design
+
+### Implementation Changes
+
+The "seams" in the P2P implementation between the higher level
+constructs (reactors), the routing layer (`Router`) and the lower
+level connection and peer management code make this operation
+relatively straight forward from an implementation perspective. The
+goal in this design is to minimize the impact to the reactors
+(potentially entirely,) and completely remove the lower level
+aspects (e.g. `Transport`, `Connection` and `PeerManager`) using the
+separation afforded by the `*Router` layer. The current state of the
+code makes these changes relatively surgical, and limited to a small
+number of methods:
+
+- `p2p.Router.OpenChannel` will still return a `Channel` structure
+  which will continue to serve as a pipe between the reactors and the
+  `Router`. The implementation will no longer need the queue
+  implementation and this operation will instead start goroutines that
+  are responsible for routing the messages from the channel to libp2p
+  fundamentals, replacing the current `p2p.Router.routeChannel`.
+
+- The current `p2p.Rotuer.dialPeers` and `p2p.Router.acceptPeers`,
+  which are responsible for establishing outbound and inbound
+  connections respectively, will be removed, along with
+  `p2p.Router.openConnection` and the libp2p connection manager will
+  be responsible for maintaining network connectivity.
+
+- The `p2p.Channel` interface should change to replace the golang
+  channels with a more functional interface for sending messages. The
+  new methods on this object, will take contexts to support safe
+  cancellation and return errors, and generally block rather than
+  return asynchronously. The `Out` channel through which
+  reactors send messages to Peers, will be replaced by a `Send`
+  method, and the Error channel will be replaced by an `Error`
+  method. 
+
+- Rectors will have access to an interface that will allow them to
+  access information from the current state of the Peer information
+  from libp2p. This will supplant the  `p2p.PeerUpdates`
+  subscription.
+
+### Upgrade and Compatibility
+
+Because the routers and all current P2P code is in the `internal`
+package and not part of the public API, the only changes to the public
+API surface area of the tendermint will be different configuration
+file options, replacing the current P2P options with options relevant
+to libp2p.
```

Comment:
```suggestion
Because the routers and all current P2P code is in `internal`
packages and not part of the public API, the only changes to the public
API surface area of Tendermint will be different configuration
file options, replacing the current P2P options with options relevant
to libp2p.
```

Maybe also we should plan to provide a tool to migrate existing settings to compatible equivalents in the new config section? Or maybe just agree to honor sensible interpretations of the old values for some period of time?